### PR TITLE
error message for TStruct.insert when inserting into missing struct with a required field

### PR DESCRIPTION
**Pull request by catoverdrive**
@danking I think this is what you meant. I looked at unsafeInserter also but I don't see where/if it tries to expand missing structs?

### CODE REVIEW

**Comment by danking** [comment]
I feel like this should bail as soon as we know the types are incompatible, rather than succeed if it turns out that there are no missing structs. @cseed agree/disagree?

**Comment by danking** [comment]
This will fail until https://github.com/hail-is/hail/pull/2425 lands.

**Comment by danking** [comment]
@catoverdrive `updateKey` is the method you're looking for in the unsafe case.

File: src/main/scala/is/hail/expr/Type.scala [hunk 1]
```diff
@@ -1882,7 +1882,7 @@ final case class TStruct(fields: IndexedSeq[Field], override val required: Boole
         } ]
            |  Hint: use `drop' or `select' to remove these fields from one side""".stripMargin)
 
-    val newStruct = TStruct(fields ++ other.fields.map(f => f.copy(index = f.index + size)))
+    val newStruct = TStruct(fields ++ other.fields.map(f => f.copy(index = f.index + size)), required || other.required)
```
Comment (danking) [inline src/main/scala/is/hail/expr/Type.scala]: I believe this modification of the `apply` method will compile:

```scala

  def apply(args: (String, Type)*): TStruct =
    apply(false, args:_*)

  def apply(required: Boolean, args: (String, Type)*): TStruct =
    ...
```

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/Type.scala]: done

File: src/main/scala/is/hail/expr/Type.scala [hunk 2]
```diff
@@ -1598,6 +1598,13 @@ object TStruct {
       .map { case ((n, t), i) => Field(n, t, i) }
       .toArray)
```
Comment (danking) [inline src/main/scala/is/hail/expr/Type.scala]: Can you call the latter from this one so we don't duplicate the code :3

```scala
apply(false, args:_*)
```

**Comment by catoverdrive** [comment]
I'm closing this and will re-open once I've rooted out the implications of this change.

**Pull request closed without merging**